### PR TITLE
[SW-79] DEX transaction – Transaction detail

### LINKED ISSUE #1457: Create Custom View TX Screen for DEX Swaps
Augment the TX view screen and probably TX list with information about the swap.

### PR DESCRIPTION
**Pull request by pstachel**
fixes #1457

### CODE REVIEW

File: src/popup/router/components/TransactionTokenRow.vue [hunk 1]
```diff
@@ -0,0 +1,81 @@
+<template>
+  <div
+    class="transaction-token-row"
+  >
```
Comment (CedrikNikita) [inline src/popup/router/components/TransactionTokenRow.vue:4]:
```suggestion
  <div class="transaction-token-row">
```

File: src/popup/router/components/TransactionTokenRow.vue [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+export default {
+  components: { Tokens, TokenAmount },
+  props: {
+    token: {
+      type: Object,
+      required: true,
+    },
+    hasError: {
+      type: Boolean,
+      default: false,
```
Comment (CedrikNikita) [inline src/popup/router/components/TransactionTokenRow.vue:33]: no need to define `default: false` for a `Boolean`

**Review by CedrikNikita — CHANGES_REQUESTED** [review]
- Incorrect information if the contract send token to your account
image

- For the remove_liquidity_ae showing a wrong information for AE (wrong symbol, wrong contract)
image

- label is wrong for the remove_liquidity, label now is: `Provide liquidity`

- PR can be rebased on develop to use `FUNCTION_TYPE_DEX` constant, if it is needed

- fix github actions warnings

**Review by CedrikNikita — COMMENTED** [review]
Should be rebased on top https://github.com/aeternity/superhero-wallet/pull/1494 PR, which is already included in the develop branch, in order to work correctly.

File: src/popup/router/components/TransactionTokenRow.vue [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    },
+    direction: {
+      type: String,
+      validator: (value) => ['sent', 'received'].includes(value),
+      default: undefined,
+    },
+  },
+  computed: {
+    ...mapState('fungibleTokens', ['availableTokens']),
```
Comment (CedrikNikita) [inline src/popup/router/components/TransactionTokenRow.vue:42]: Doesnt seems to be used in this file.

**Comment by CedrikNikita** [comment]
Still wrong information 
image
The transaction is actually a `+6 WTT`
image

https://explorer.aeternity.io/transactions/th_ahNbwrdfDDeCaSrSBcR2MPnN1f2fRRUpFgEKhXHEwpfdWzXPr

File: src/popup/router/components/TransactionDetailsPoolTokens.vue [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+      type: String,
+      required: true,
+    },
+    withAmount: {
+      type: Boolean,
+      default: true,
+    },
+  },
+  methods: {
+    ...mapActions('fungibleTokens', ['getContractTokenPairs']),
```
Comment (CedrikNikita) [inline src/popup/router/components/TransactionDetailsPoolTokens.vue:63]: Is it used in this file?

File: src/popup/router/pages/TransactionDetails.vue [hunk 5]
```diff
@@ -198,26 +230,74 @@ export default {
...
+    },
+    hasError() {
+      return this.transaction?.tx.result === 'abort';
+    },
   },
   async mounted() {
     this.transaction = this.getTx(this.hash);
     if (!this.transaction || this.transaction?.incomplete) {
       await this.$watchUntilTruly(() => this.$store.state.middleware);
+      await this.$watchUntilTruly(() => this.$store.state.sdk);
```
Comment (CedrikNikita) [inline src/popup/router/pages/TransactionDetails.vue:276]: Why do we need to wait until sdk trully here?

File: src/store/plugins/fungibleTokens.js [hunk 6]
```diff
@@ -158,6 +159,29 @@ export default (store) => {
           decodedResult !== undefined ? 'change_allowance' : 'create_allowance'
         ](activeNetwork.tipContractV2.replace('ct_', 'ak_'), allowanceAmount);
       },
+      async getContractTokenPairs(
```
Comment (CedrikNikita) [inline src/store/plugins/fungibleTokens.js:162]: This function is updated in other PR https://github.com/aeternity/superhero-wallet/pull/1493.

File: src/popup/router/components/TransactionOverview.vue [hunk 7]
```diff
@@ -16,10 +17,11 @@ export default {
   components: { Overview },
   props: {
     tx: { type: Object, required: true },
+    isDex: { type: Boolean, default: false },
```
Comment (CedrikNikita) [inline src/popup/router/components/TransactionOverview.vue:20]:
```suggestion
    isDex: { type: Boolean },
```

File: src/popup/router/components/TransactionRevertedStatus.vue [hunk 8]
```diff
@@ -0,0 +1,85 @@
+<template>
+  <div class="transaction-reverted-status">
+    <div class="status-wrapper">
+      <div class="text-wrapper">
+        <Revert class="revert-icon" />
```
Comment (CedrikNikita) [inline src/popup/router/components/TransactionRevertedStatus.vue:5]: Icon for the `returnType === 'abort'` is missing.

**Review by CedrikNikita — COMMENTED** [review]
- Create token allowance transaction look like that for me:
image
https://explorer.testnet.aeternity.io/transactions/th_2r4EQkq4ehwFt9yEWscZdH3LhytDmMctHmMohvgMoGxG7Wp8Hi
- Token icon doesn't look like that:
image
It should look like this
https://avatars.z52da5wt.xyz/ct_2kJUXqBsqeQgx5thvKpoTwhLbn95VSRRWsU1sN8Pejzd33F3XM
There is no need to show 0 after the last digit, the amount i've send was 0.04
- We should show `Superhero Dex` only for the Dex related transactions, i think contractCalls that using `router`s addreses only
image

File: src/icons/revert.svg [hunk 9]
```diff
@@ -0,0 +1,4 @@
+<svg viewBox="0 0 21 20" fill="none" xmlns="http://www.w3.org/2000/svg">
```
Comment (CedrikNikita) [inline src/icons/revert.svg:1]: Same icon was added it in the transaction list with a different name, you can reuse it.

File: src/popup/router/components/AddressShortening.vue [hunk 10]
```diff
@@ -58,22 +58,27 @@ export default {
 
 <style lang="scss" scoped>
 @use '../../../styles/variables';
+@use '../../../styles/typography';
```
Comment (CedrikNikita) [inline src/popup/router/components/AddressShortening.vue:61]: You need to make sure that in all places where it is used everything looks good.

**Review by CedrikNikita — COMMENTED** [review]
In order to be ready the swapRoute component should be exactly like in the designs (it was updated).

File: src/popup/router/components/TransactionDetailsPoolTokens.vue (hunk 4 above)
Comment (CedrikNikita) [inline src/popup/router/components/TransactionDetailsPoolTokens.vue:63]: Not used in this file.

File: src/popup/router/pages/TransactionDetails.vue [hunk 11]
```diff
@@ -201,26 +228,74 @@ export default {
...
+    Promise.all(this.tokens().map(
+      async (token) => ({ ...token, tokens: await this.prepareTokenPair(token) }),
+    )).then((values) => {
+      this.tokenList = values.filter((token) => !Number.isNaN(Number(token.amount)));
+    });
+  },
+  methods: {
+    ...mapActions('fungibleTokens', ['getContractTokenPairs']),
+    aettosToAe,
+    async prepareTokenPair(token) {
```
Comment (CedrikNikita) [inline src/popup/router/pages/TransactionDetails.vue:286]: All the needed information is in the resolvers output, no need to use getContractTokenPairs.

File: src/popup/router/components/TransactionTokenRow.vue [hunk 12]
```diff
@@ -0,0 +1,86 @@
+<template>
+  <div
+    class="transaction-token-row"
+  >
```
Comment (CedrikNikita) [inline src/popup/router/components/TransactionTokenRow.vue:4]:
```suggestion
  <div class="transaction-token-row">
```

File: src/popup/router/pages/TransactionDetails.vue (hunk 5 above)
Comment (CedrikNikita) [reply src/popup/router/pages/TransactionDetails.vue:274]: We don't actually need to use this function, so we can omit waiting for sdk.

File: src/popup/utils/transactionTokenInfoResolvers.js [hunk 13]
```diff
@@ -28,6 +28,10 @@ const genLiquiditySwapResolver = (
     const symbolB = tokenB.isAe ? 'AE' : tokens?.[tokenB.contractId].symbol;
     poolTokenSymbol = `${symbolA}/${symbolB}`;
```
Comment (CedrikNikita) [inline src/popup/utils/transactionTokenInfoResolvers.js:29]: All the changes in that file are obviously a rebase mistake.

**Review by CedrikNikita — CHANGES_REQUESTED** [review]
SwapRoute component:
- There should be no line in the end
- First token to show is WAE, there is no pool fee in WAE contract (double check it with Bogdan)
- There is no need to show the pool info in the swapRoutes, only the direct swap (check once again the new designs)

image

We can extract the rows of a token from transaction list and transaction details.

File: src/icons/failed.svg [hunk 14]
```diff
@@ -0,0 +1,5 @@
+<svg viewBox="0 0 21 21" fill="none" xmlns="http://www.w3.org/2000/svg">
```
Comment (CedrikNikita) [inline src/icons/failed.svg:1]: There is a warning svg that is almost identical, we spoke with Tsvetan you should use the new one everywhere now.

File: src/popup/router/pages/TransactionDetails.vue [hunk 15]
```diff
@@ -4,151 +4,165 @@
...
-            slot="value"
-            :amount="transaction.tx.gasPrice"
-            symbol="ættos"
-            hide-fiat
+        </template>
+      </Plate>
+      <div class="content">
+        <TransactionOverview v-bind="transaction" />
+        <div class="data-grid">
+          <template v-if="isSwap && !hasError && UNFINISHED_FEATURES">
```
Comment (CedrikNikita) [inline src/popup/router/pages/TransactionDetails.vue:29]:
```suggestion
          <template v-if="isSwap && !hasError">
```

**Review by CedrikNikita — COMMENTED** [review]
@paolomolo is the way it is hown something we agreed on?
image

- i don't think that layout should jump here

Copy hash

**Comment by CedrikNikita** [comment]
image

**Comment by CedrikNikita** [comment]
No contract
In the transaction list we have visualy more info on reverted remove liquidity transaction

File: src/popup/router/components/Overview.vue [hunk 16]
```diff
@@ -176,23 +220,32 @@ export default {
...
+      @extend %face-mono-12-medium;
+
       height: 48px;
     }
   }
+
+  .truncated-string {
+    background: red;
+    text-decoration: underline;
+  }
```
Comment (ifaouibadi) [inline src/popup/router/components/Overview.vue:249]: not used anywhere in the app

File: src/popup/router/pages/TransactionDetails.vue [hunk 17]
```diff
@@ -4,151 +4,165 @@
...
+          <DetailsItem
+            v-if="tipUrl"
+            :label="$t('pages.transactionDetails.tipUrl')"
+            class="tip-url"
+            data-cy="tip-url"
+          >
+            <CopyButton
+              slot="label"
+              :value="tipUrl"
+              message="URL copied"
```
Comment (ifaouibadi) [inline src/popup/router/pages/TransactionDetails.vue:54]: URL copied should come from translations

**Review by ifaouibadi — CHANGES_REQUESTED** [review]

File: src/popup/router/components/Overview.vue [hunk 18]
```diff
@@ -90,18 +109,36 @@ export default {
...
+    txFunction: { type: String, default: null },
     sender: { type: Object, required: true },
     recipient: { type: Object, required: true },
   },
+  computed: {
+    isNewUI() {
+      return !!this.$route.meta.newUI;
+    },
+    getTxType() {
+      return this.$te(`transaction.dexType.${getDexTransactionTag[this.txFunction] || this.txFunction}`)
```
Comment (ifaouibadi) [inline src/popup/router/components/Overview.vue:125]:
```suggestion
      return this.$t(`transaction.dexType.${getDexTransactionTag[this.txFunction] || this.txFunction}`)
```

**Review by ifaouibadi — CHANGES_REQUESTED** [review]

File: src/popup/router/components/Overview.vue (hunk 18 above)
Comment (pstachel) [reply src/popup/router/components/Overview.vue:125]: It's not the same, `$te` checks whether translation exists - if we will proceed with your suggestion & there won't be exact translation we will end up with always being true which will cause some problems with display

**Comment by CedrikNikita** [comment]
Adjustments:
- **fix** wrong condition for the pool fee in swap route
- **fix**: swapRoute component did not  show hidden AE to WAE swap
- **fix**: swapRoute was working completely different, showing pairs instead of tokens that being swap
- **fix**: swapRoute **did not include the route itself (wasnt showing actual route)**
- **fix**: there was no AE icon for sendTx transactions
- **fix**: there was no icon for poolToken in reverted transactions

SwapRoute:
|type|before|after|
|---|---|---|
|Swap token for **AE**|image|image|
|Swap **AE** for token|image|image|
|Swap **WAE** for token|image|image|
|Swap token for **WAE** |image|image|
|Swap token for token (**without** complex route) |image|image|
|Swap token for token (**with** complex route)|image|image|

**Review by CedrikNikita — APPROVED** [review]
Looks good from code perspective.

**Pull request merged**